core/src/cpu/air/ecall.rs

```diff
@@ -0,0 +1,224 @@
+use p3_air::AirBuilder;
+use p3_field::AbstractField;
+
+use crate::air::{BaseAirBuilder, WordAirBuilder};
+use crate::cpu::air::{Word, POSEIDON_NUM_WORDS, PV_DIGEST_NUM_WORDS};
+use crate::cpu::columns::{CpuCols, OpcodeSelectorCols};
+use crate::memory::MemoryCols;
+use crate::operations::IsZeroOperation;
+use crate::runtime::SyscallCode;
+use crate::stark::{CpuChip, SP1AirBuilder};
+
+impl CpuChip {
+    /// Whether the instruction is an ECALL instruction.
+    pub(crate) fn is_ecall_instruction<AB: SP1AirBuilder>(
+        &self,
+        opcode_selectors: &OpcodeSelectorCols<AB::Var>,
+    ) -> AB::Expr {
+        opcode_selectors.is_ecall.into()
+    }
+
+    /// Constraints related to the ECALL opcode.
+    pub(crate) fn ecall_eval<AB: SP1AirBuilder>(
+        &self,
+        builder: &mut AB,
+        local: &CpuCols<AB::Var>,
+    ) -> (AB::Expr, AB::Expr, AB::Expr, AB::Expr) {
+        let ecall_cols = local.opcode_specific_columns.ecall();
+        let is_ecall_instruction = self.is_ecall_instruction::<AB>(&local.selectors);
+        // The syscall code is the read-in value of op_a at the start of the instruction.
+        let syscall_code = local.op_a_access.prev_value();
+        // We interpret the syscall_code as little-endian bytes and interpret each byte as a u8
+        // with different information. Read more about the format in runtime::syscall::SyscallCode.
+        let syscall_id = syscall_code[0];
+        let send_to_table = syscall_code[1]; // Does the syscall have a table that should be sent.
+        let num_cycles = syscall_code[2]; // How many extra cycles to increment the clk for the syscall.
+
+        // Check that the ecall_mul_send_to_table column is equal to send_to_table * is_ecall_instruction.
+        // This is a separate column because it is used as a multiplicity in an interaction which
+        // requires degree 1 columns.
+        builder
+            .when(is_ecall_instruction.clone())
+            .assert_eq(send_to_table, local.ecall_mul_send_to_table);
+        builder.send_syscall(
+            local.shard,
+            local.clk,
+            syscall_id,
+            local.op_b_val().reduce::<AB>(),
+            local.op_c_val().reduce::<AB>(),
+            local.ecall_mul_send_to_table,
+        );
+
+        // Constrain EcallCols.is_enter_unconstrained.result == syscall_id is ENTER_UNCONSTRAINED.
```

Comment:
Not sure if this return pattern is really getting you anything, maybe we should just use ecall_cols directly?